pkg/collector/py/datadog_agent.c

```diff
@@ -101,12 +110,127 @@ static PyObject *get_subprocess_output(PyObject *self, PyObject *args) {
     return py_result;
 }
 
+static PyObject *add_external_tags(PyObject *self, PyObject *args) {
+    PyObject *input_list = NULL;
+    PyGILState_STATE gstate = PyGILState_Ensure();
+
+    // function expects only one positional arg containing a list
+    if (!PyArg_ParseTuple(args, "O", &input_list)) {
+        PyGILState_Release(gstate);
+        return NULL;
+    }
+
+    // if not a list, set an error
+    if (!PyList_Check(input_list)) {
+        PyErr_SetString(PyExc_TypeError, "function arg must be a list");
+        PyGILState_Release(gstate);
+        return NULL;
+    }
+
+    // if the list is empty do nothing
+    int input_len = PyList_Size(input_list);
+    if (input_len == 0) {
+        PyGILState_Release(gstate);
+        Py_RETURN_NONE;
+    }
+
+    int i;
+    for (i=0; i<input_len; i++) {
+        PyObject *tuple = PyList_GetItem(input_list, i);
+
+        // list must contain only tuples in form ('hostname', {'source_type': ['tag1', 'tag2']},)
+        if (!PyTuple_Check(tuple)) {
+            PyErr_SetString(PyExc_TypeError, "external host tags list must contain only tuples");
+            PyGILState_Release(gstate);
+            return NULL;
+        }
+
+        // first elem is the hostname
+        const char *hostname = PyString_AsString(PyTuple_GetItem(tuple, 0));
+        // second is a dictionary
+        PyObject *dict = PyTuple_GetItem(tuple, 1);
+        if (!PyDict_Check(dict)) {
+            PyErr_SetString(PyExc_TypeError, "second elem of the host tags tuple must be a dict");
+            PyGILState_Release(gstate);
+            return NULL;
+        }
+
+        // dict contains only 1 key, if dict is empty don't do anything
+        Py_ssize_t pos = 0;
+        PyObject *key = NULL, *value = NULL;
+        if (!PyDict_Next(dict, &pos, &key, &value)) {
+            continue;
+        }
+
+        // key is the source type (e.g. 'vsphere') value is the list of tags
+        const char *source_type = PyString_AsString(key);
+        if (!PyList_Check(value)) {
+            PyErr_SetString(PyExc_TypeError, "dict value must be a list of tags ");
+            PyGILState_Release(gstate);
+            return NULL;
+        }
+
+        // allocate an array of char* to store the tags we'll send to the Go function
+        char **tags;
+        int tags_len = PyList_Size(value);
+        if(!(tags = (char **)malloc(sizeof(char *)*tags_len))) {
+            PyErr_SetString(PyExc_MemoryError, "unable to allocate memory, bailing out");
+            PyGILState_Release(gstate);
+            return NULL;
+        }
+
+        // copy the list of tags into an array of char*
+        int j, actual_size = 0;
+        for (j=0; j<tags_len; j++) {
+            PyObject *s = PyList_GetItem(value, j);
+            if (s == NULL) {
+                continue;
+            }
+
+            char *tag = PyString_AsString(s);
+            if (tag == NULL) {
+                continue;
+            }
+
```

Comment:
In line 103 above, you allocate an array of char*, and then use the result of PyString_AsString() directly. But here  you go through the work of malloc()ing memory and copying.  The `PyString_AsString()` docs don't indicate how long you can trust the buffer that's returned to be valid;  is the code in `get_subprocess_output()` insufficient?  Or is this code overkill?

Reply:
`PyString_AsString` gives you a "view" of the internal buffer stored in the PyObject, thus the lifetime is the same as the object on the Python side. In this case I access the buffer within a `for` loop, meaning that potentially, once you exit the loop the buffer might be gone and you pass around a pointer to invalid memory. 
I though a copy required less code, being the alternative to increment the reference counter for each PyObject contained in the list, doing whatever with the `char*` pointing to their internal buffer (in this case invoking the Go func) then decrement the reference counter and let Python clean up the mess.
Maybe it wouldn't be that much more code though 🤔

Reply:
I think I'm going to keep it like this, explicit deallocation seems more readable